### ISSUE TITLE
False alarm with derive_ord_xor_partial_ord

### ISSUE THREAD

**Issue by kangalioo** [rust-lang/rust-clippy#6219]
In my application, I have a struct wrapping a `f32`, but guaranteeing non-NaN values. Therefore, this struct can have an Ord implementation.

I've implemented this using:
```rust
#[derive(PartialOrd)]
struct MyWrapper {
    inner: f32,
}

impl Ord for MyWrapper {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        other.partial_cmp(other).expect("Can't happen; this wrapper guarantees non-NaN")
    }
}
```

This triggers a deny-by-default warning, telling me that I'm making my Ord and PartialOrd implementation not match ([link](https://rust-lang.github.io/rust-clippy/master/index.html#derive_ord_xor_partial_ord)). However, as you can see, this is not the case here.

This false alarm was easily fixed with `#[allow(clippy::derive_ord_xor_partial_ord)]`, however it may be good to
- either: add this case to the "Known problems" list in the Clippy lint index
- or: improve the lint trigger detection to check whether the Ord implementation internally calls the PartialCmp implementation, and if yes, suppress the lint

**Comment by giraffate** [comment]
Repro: [playground](https://play.rust-lang.org/?version=beta&mode=debug&edition=2018&gist=795aa41345c637200d89e17b133ad2b3)

@rustbot modify lables: +L-bug

**Comment by wfraser** [comment]
Seems incredibly low risk if one of the operations is implemented in terms of the other.

Looking at the list of crates that have ignored this lint, that reasoning covers most of them. Several of them are also non-NaN float wrappers as well.

https://dtolnay.github.io/noisy-clippy/derive_ord_xor_partial_ord.html#local

In fact, I would argue the chances of getting it wrong are *higher* if someone manually implements both operations. *That* is what should be linted against: if you have explicit impls of both and one is *not* implemented in terms of the other.

**Comment by spearman** [comment]
I have the same false positive for Normalized(f32) type that guarantees the contained value to be in the range 0.0-1.0
